include/alpaka/atomic/AtomicOmpBuiltIn.hpp

```diff
@@ -0,0 +1,326 @@
+* (at your option) any later version.
+*
+* alpaka is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with alpaka.
+* If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#ifdef _OPENMP
+
+#include <alpaka/atomic/Traits.hpp>
+
+namespace alpaka
+{
+    namespace atomic
+    {
+        //#############################################################################
+        //! The OpenMP accelerator atomic ops.
+        //
+        //  Atomics can be used in the blocks and threads hierarchy levels.
+        //  Atomics are not guaranteed to be save between devices or grids.
+        class AtomicOmpBuiltIn
+        {
+        public:
+            using AtomicBase = AtomicOmpBuiltIn;
+
+            //-----------------------------------------------------------------------------
+            AtomicOmpBuiltIn() = default;
+            //-----------------------------------------------------------------------------
+            ALPAKA_FN_HOST AtomicOmpBuiltIn(AtomicOmpBuiltIn const &) = delete;
+            //-----------------------------------------------------------------------------
+            ALPAKA_FN_HOST AtomicOmpBuiltIn(AtomicOmpBuiltIn &&) = delete;
+            //-----------------------------------------------------------------------------
+            ALPAKA_FN_HOST auto operator=(AtomicOmpBuiltIn const &) -> AtomicOmpBuiltIn & = delete;
+            //-----------------------------------------------------------------------------
+            ALPAKA_FN_HOST auto operator=(AtomicOmpBuiltIn &&) -> AtomicOmpBuiltIn & = delete;
+            //-----------------------------------------------------------------------------
+            /*virtual*/ ~AtomicOmpBuiltIn() = default;
+        };
+
+        namespace traits
+        {
+            //#############################################################################
+            //! The OpenMP accelerator atomic operation: ADD
+            template<
+                typename T,
+                typename THierarchy>
+            struct AtomicOp<
+                op::Add,
+                atomic::AtomicOmpBuiltIn,
+                T,
+                THierarchy>
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_ACC_CUDA_ONLY static auto atomicOp(
+                    atomic::AtomicOmpBuiltIn const &,
+                    T * const addr,
+                    T const & value)
+                -> T
+                {
+                    T old;
+                    auto & ref(*addr);
+                    // atomically update ref, but capture original value in old
+                    #pragma omp atomic capture
```

Comment:
`capture` is not available in OpenMP 2.

Reply:
I will guard all Spezialisation for OpenMP2. I need to check first since which version capture is supported.